src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp

```diff
@@ -385,13 +387,83 @@ namespace OpenLoco::World::MapGenerator
         }
     }
 
+    static void updateProgress(uint8_t value);
+
+    // 0x004595B7
+    static bool isIndustryCargoAvailable(uint8_t cargoType)
+    {
+        registers regs;
+        regs.eax = cargoType;
+        return call(0x004595B7, regs) & X86_FLAG_CARRY;
+    }
+
     // 0x004597FD
     static void generateIndustries(uint32_t minProgress, uint32_t maxProgress)
     {
-        registers regs;
-        regs.eax = minProgress;
-        regs.ebx = maxProgress;
-        call(0x004597FD, regs);
+        auto numIndustriesAvailable = 0;
+        for (auto indObjId = 0U; indObjId < ObjectManager::getMaxObjects(ObjectType::industry); indObjId++)
+        {
+            if (ObjectManager::get<IndustryObject>(indObjId) != nullptr)
+            {
+                numIndustriesAvailable++;
+            }
+        }
+
+        CompanyManager::setUpdatingCompanyId(CompanyId::neutral);
+
+        auto progressTicksPerIndustry = (maxProgress - minProgress) / numIndustriesAvailable;
+        auto currentProgress = minProgress;
+
+        for (auto indObjId = 0U; indObjId < ObjectManager::getMaxObjects(ObjectType::industry); indObjId++)
+        {
+            auto* industryObj = ObjectManager::get<IndustryObject>(indObjId);
+            if (industryObj == nullptr)
+            {
+                continue;
+            }
+
+            currentProgress += progressTicksPerIndustry;
+            updateProgress(currentProgress);
+
+            // Check if industry is available at present
+            if (getCurrentYear() < industryObj->designedYear || getCurrentYear() > industryObj->obsoleteYear)
+            {
+                continue;
+            }
+
+            // Check if required cargo is available
+            if (industryObj->requiredCargoType[0] != 0xFF)
+            {
+                auto numCargoSpecified = 0;
+                auto numCargoAvailable = 0;
+                for (auto cargoType : industryObj->requiredCargoType)
+                {
+                    if (cargoType == 0xFF)
+                        continue;
+
+                    numCargoSpecified++;
+                    if (isIndustryCargoAvailable(cargoType))
+                        numCargoAvailable++;
+                }
+
+                if (industryObj->hasFlags(IndustryObjectFlags::requiresAllCargo))
+                {
+                    if (numCargoSpecified != numCargoAvailable)
+                        continue;
+                }
+                else if (numCargoAvailable == 0)
+                {
+                    continue;
+                }
+            }
+
+            const uint8_t numIndustriesFactor = S5::getOptions().numberOfIndustries;
+            const auto numIndustriesToCreate = IndustryManager::capOfTypeOfIndustry(indObjId, numIndustriesFactor);
```

Comment:
This value is off. It seems to be half the size of what vanilla produces. Have not been able to narrow down the cause yet.

Reply:
Turns out I have been chasing ghosts here. For a medium number of industries, master generates around 40 industries, and 60 industries for a high number. We're seeing similar numbers on this PR.

For some reason, I was expecting these numbers to be way closer to the kMaxIndustries limit, 128. I clearly thought I had seen this on master, but evidently, this is not the case.